encoding/thrift/internal/observabilitytest/test/testserviceserver/server.go

```diff
@@ -104,3 +107,50 @@ func (h handler) Call(ctx context.Context, body wire.Value) (thrift.Response, er
 
 	return response, err
 }
+
+type Call_NoWireHandler struct{ impl Interface }
+
+func (h Call_NoWireHandler) Handle(ctx context.Context, nwc *thrift.NoWireCall) (thrift.NoWireResponse, error) {
+	var (
+		args test.TestService_Call_Args
+
+		rw stream.ResponseWriter
+
+		err error
+	)
+
+	if nwc.RequestReader != nil {
+
+		rw, err = nwc.RequestReader.ReadRequest(ctx, nwc.EnvelopeType, nwc.Reader, &args)
+
+	} else {
+		err = args.Decode(nwc.StreamReader)
+	}
```

Comment:
Commenting on this generated code (I know the code was added in encoding/thrift/thriftrw-plugin-yarpc/server.go, but easier to understand what is going on).

Why do we have two ways for decoding the request here? How does it happen?

Reply:
This happens because the caller of this `Handle` may not have a `proto` that implements a `stream.RequestReader`.  If the caller has a `proto` that implements `stream.RequestReader`, then the logic is quite simple since at that call site, it simply returns that request:
```
	if reqReader, ok := proto.(stream.RequestReader); ok {
		nwc := NoWireCall{
			Reader:        treq.Body,
			RequestReader: reqReader,
			EnvelopeType:  reqEnvelopeType,
		}
		return noWireHandler.Handle(ctx, &nwc)
	}
```

The differences between this and the Wire utilization, is that even outside of being able to make the request to `Handle`, `Read...` operations are still needed once the request is processed.

An argument can be made that will wrap a `stream.Protocol` into a struct that implements a `stream.RequestReader` which should simplify this.   That should make this a lot cleaner - will implement that.